File: generate_index.py
import os
import json
import re

NOTES_DIR = "notes"
OUTPUT_FILE = "notes_index.json"

def extract_title(path, filename):
    """Extract first H1 (# ...) line or fallback to filename."""
    try:
        with open(path, encoding="utf-8") as f:
            for line in f:
                match = re.match(r"#\s+(.+)", line.strip())
                if match:
                    return match.group(1).strip()
    except Exception:
        pass
    return filename

def generate_index():
    files = [f for f in os.listdir(NOTES_DIR) if f.endswith(".md")]
    files.sort()

    data = []
    for f in files:
        fullpath = os.path.join(NOTES_DIR, f)
        title = extract_title(fullpath, f)
        data.append({"file": f, "title": title})

    with open(OUTPUT_FILE, "w", encoding="utf-8") as out:
        json.dump(data, out, indent=2, ensure_ascii=False)

    print(f"✅ Generated {OUTPUT_FILE} with {len(files)} notes.")

if __name__ == "__main__":
    generate_index()

File: generate_index.js
const fs = require("fs");
const path = require("path");
const jsyaml = require("js-yaml");

const NOTES_DIR = path.join(__dirname, "notes");
const OUTPUT_FILE = path.join(__dirname, "notes_index.json");

const files = fs.readdirSync(NOTES_DIR).filter(f => f.endsWith(".md"));

const notes = [];

for (const file of files) {
  const fullPath = path.join(NOTES_DIR, file);
  let content = fs.readFileSync(fullPath, "utf-8");

  let frontmatter = {};
  if (content.startsWith("---")) {
    const end = content.indexOf("---", 3);
    if (end !== -1) {
      const yamlText = content.slice(3, end).trim();
      frontmatter = jsyaml.load(yamlText) || {};
      content = content.slice(end + 3).trim();
    }
  }

  notes.push({
    file,
    title: frontmatter.title || file.replace(".md", ""),
    content,
    frontmatter
  });
}

fs.writeFileSync(OUTPUT_FILE, JSON.stringify(notes, null, 2), "utf-8");
console.log("notes_index.json generated!");
